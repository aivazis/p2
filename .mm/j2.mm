# -*- Makefile -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2019 all rights reserved

# project meta-data
j2.major := $(repo.major)
j2.minor := $(repo.minor)
j2.micro := $(repo.micro)
j2.revision := $(repo.revision)

# j2 builds a python package
j2.packages :=
# libraries
j2.libraries := j2.lib
# python extensions
j2.extensions :=
# and test suites
j2.tests := j2.lib.tests

# the j2 library settings
j2.lib.stem := j2
j2.lib.c++.defines += PYRE_CORE
j2.lib.c++.flags += $($(compiler.c++).std.c++17)

# get the testsuites
include j2.tests

# end of file
