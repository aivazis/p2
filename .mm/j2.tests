# -*- Makefile -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2019 all rights reserved


# global testsuite settings
j2.lib.tests.harness :=
j2.lib.tests.flags += -Wall $($(compiler.c++).std.c++17)
j2.lib.tests.defines :=

# end of file
