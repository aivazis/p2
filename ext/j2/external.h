// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>

// pybind support
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/stl_bind.h>

// make certain STL containers opaque
PYBIND11_MAKE_OPAQUE(pyre::journal::page_t);
PYBIND11_MAKE_OPAQUE(pyre::journal::notes_t);


// type aliases
namespace pyre::journal::py {
    // import {pybind11}
    namespace py = pybind11;
    // get the special {pybind11} literals
    using namespace py::literals;
}


// end of file
