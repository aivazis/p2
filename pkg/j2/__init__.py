# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# administrative
def copyright():
    """
    Return the pyre copyright note
    """
    return print(meta.header)


def license():
    """
    Print the pyre license
    """
    # print it
    return print(meta.license)


def version():
    """
    Return the pyre version
    """
    return meta.version


def credits():
    """
    Print the acknowledgments
    """
    # print it
    return print(meta.acknowledgments)


# publish
# the metadata
from . import meta

# set up a marker about whether we are going to load and publish the bindings
without_libjournal = False

# get the {__main__} module
import __main__
# check whether
try:
    # the user has expressed an opinion
    without_libjournal = __main__.journal_no_libjournal
# if not
except AttributeError:
    # no worries
    pass

# if we are allowed
if without_libjournal is False:
    # load the bindings
    from .ext import libjournal
    # if something went wrong
    if libjournal is None:
        # indicate that we don't have access to the binding
        without_libjournal = True

# if there is no functional extension module
if without_libjournal:
    # fall back on the pure python implementation
    # the devices
    from .Trash import Trash as trash
    from .Stream import Stream as stream
    from .Console import Console as cout
    from .ErrorConsole import ErrorConsole as cerr

    # the channels
    from .Null import Null as null

    from .Debug import Debug as debug
    from .Firewall import Firewall as firewall

    from .Informational import Informational as info
    from .Warning import Warning as warning
    from .Error import Error as error

    # get the singleton with the global state
    from .Chronicler import Chronicler as chronicler
    # attach the default device
    chronicler().device = cout()

# otherwise
else:
    # minimal configuration, for now
    # get the console
    from .Console import Console as cout
    # get the singleton with the global state
    from .Chronicler import Chronicler as chronicler
    # attach the default device
    chronicler().device = cout()


# end of file
