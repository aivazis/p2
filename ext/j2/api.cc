// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// add bindings to the inventory
void
p2::libjournal::
api(py::module & m) {
    // easy access to the manager of the global state
    m.attr("chronicler") = m.attr("Chronicler");

    // registration of the application name
    m.def("application",
          // the implementation
          &pyre::journal::application, "name"_a,
          // the docstring
          "register the application {name}"
          );

    // all done
    return;
}


// end of file
