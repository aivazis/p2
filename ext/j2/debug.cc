// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// add bindings for the debug channel
void
p2::libjournal::
debug(py::module & m) {

    // type aliases for the member functions (mfp: method pointer)
    // verbosity
    using getVerbosity_mfp = debug_t::verbosity_type (debug_t::*)() const;
    using setVerbosity_mfp = debug_t::verbosity_type (debug_t::*)(debug_t::verbosity_type);
    // state
    using getState_mfp = debug_t::state_type (debug_t::*)() const;
    using setState_mfp = debug_t::state_type (debug_t::*)(debug_t::state_type);
    // device
    using getDevice_mfp = debug_t::device_pointer (debug_t::*)() const;
    using setDevice_mfp = debug_t::device_pointer (debug_t::*)(debug_t::device_pointer);
    // metadata
    using getMetadata_mfp = debug_t::metadata_type & (debug_t::*)() const;


    // the debug channel interface
    py::class_<debug_t>(m, "Debug")
        // the constructor
        .def(py::init<const string_t &>(), "name"_a)

        // accessors
        // the name; read-only property
        .def_property_readonly("name", &debug_t::name)

        // the verbosity level
        .def_property("verbosity",
                      // the getter
                      (getVerbosity_mfp) &debug_t::verbosity,
                      // the setter
                      (setVerbosity_mfp) &debug_t::verbosity,
                      // the docstring
                      "access the verbosity level"
                      )

        // the channel state; mutable property
        .def_property("state",
                      // the getter
                      (getState_mfp) &debug_t::state,
                      // the setter
                      (setState_mfp) &debug_t::state,
                      // the docstring
                      "access the channel activation state"
                      )

        // the registered device: mutable property
        .def_property("device",
                      // the getter
                      (getDevice_mfp) &debug_t::device,
                      // the setter
                      (setDevice_mfp) &debug_t::device,
                      // the docstring
                      "access the output device"
                      )

        // the registered device: read-only property
        .def_property_readonly("meta",
                               // the getter
                               (getMetadata_mfp) &debug_t::metadata,
                               // the docstring
                               "access the channel metadata"
                      )

        // static interface
        // the default state
        .def_static("defaultState", &debug_t::defaultState)
        // the default device
        .def_static("defaultDevice", (debug_t::device_pointer (*)()) &debug_t::defaultDevice)
        // done
        ;

    // all done
    return;
}


// end of file
