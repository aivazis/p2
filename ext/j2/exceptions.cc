// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// add bindings to the inventory
void
pyre::journal::py::
exceptions(py::module & m) {
    // when {debug} channels are fatal
    py::register_exception<debug_error>(m, "DebugError");
    // when {firewalls} are fatal
    py::register_exception<firewall_error>(m, "FirewallError");
    // when user facing channels are fatal
    py::register_exception<application_error>(m, "ApplicationError");

    // all done
    return;
}


// end of file
