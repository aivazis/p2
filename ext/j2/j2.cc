// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// external dependencies
#include "external.h"
// namespace setup
#include "forward.h"


// the module entry point
PYBIND11_MODULE(j2, m) {
    // the doc string
    m.doc() = "the journal extension module";

    // bind the opaque types
    p2::libjournal::opaque(m);
    // global state
    p2::libjournal::chronicler(m);
    // devices
    p2::libjournal::devices(m);
    // channels
    p2::libjournal::debug(m);
}


// end of file
