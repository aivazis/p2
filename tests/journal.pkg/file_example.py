#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


def test():
    """
    Verify that we can build a stream based device
    """
    # pick a file name
    filename = "file_example.out"
    # make a stream
    ostr = open(filename, mode="w")

    # get the package
    import j2
    # turn it into a journal device
    device = j2.stream(name=filename, stream=ostr)

    # make a debug channel
    channel = j2.debug(name="tests.journal.debug")
    # activate it
    channel.activate()
    # but send the output to trash
    channel.device = device

    # add some metadats
    channel.meta["time"] = "now"
    # inject
    channel.line("debug channel:")
    channel.log("    hello world!")

    # get os services
    import os
    # so we can remove the file
    os.unlink(filename)

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
