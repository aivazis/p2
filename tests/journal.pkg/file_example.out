/Users/mga/dv/p2/tests/journal.pkg/file_example.py:36:test:
 >> tests.journal.debug(debug)
 >> application: journal
 >> time: now
 -- debug channel:
 --     hello world!
